### PR TITLE
Add alexaSmartHome event

### LINKED ISSUE #3955: Separate Event for Alexa Smart Home Skill Trigger
# This is a Feature Proposal

## Description
AWS offers a separate trigger called 'Alexa Smart Home' for their Smart Home skills, besides the regular 'Alexa Skills Kit' trigger. As these both work in different ways, Serverless's current 'alexaSkill' event does not work for Smart Home skills. Adding a separate event type for this trigger allows the Serverless framework to work for Smart Home skills, which it currently doesn't.

Similar or dependent issues:
* #1409 

## Additional Data
Serverless Framework Version: latest

### PR DESCRIPTION
**Pull request by horike37**
## What did you implement:
Added `alexaSmartHome` event as a new one.

Closes #3955

## How did you implement it:
I have created `alexaSmartHome` directory under the `events` one and implemented it.
https://github.com/serverless/serverless/tree/master/lib/plugins/aws/package/compile/events

## How can we verify it:
`amzn1.ask.skill.xx-xx-xx-xx` is an application ID for Alexa Smart Home Skill.
First, please sign up [Amazon Developer  Console](https://developer.amazon.com/) and get your application ID.
Then, write up a serverless.yml as follow, and run `sls deploy`

```yml
functions:
  mySkill:
    handler: mySkill.handler
    events:
      - alexaSmartHome: amzn1.ask.skill.xx-xx-xx-xx
```

or 

```yml
functions:
  mySkill:
    handler: mySkill.handler
    events:
      - alexaSmartHome: 
           appId: amzn1.ask.skill.xx-xx-xx-xx
           enabled: <true or false>
```

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by pmuens** [comment]
Great! Thanks for PRing this @horike37 👍 🎉

We'll review this ASAP 💪

**Comment by pmuens** [comment]
Good catch @jvsinclair 👍

Yes, I agree. Adding a section in the docs should be sufficient here (and will avoid confusion).

@horike37 let us know if you need any help with that!

File: docs/providers/aws/events/alexa-smart-home.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+<!-- DOCS-SITE-LINK:START automatically generated  -->
+### [Read this on the main serverless docs site](https://www.serverless.com/framework/docs/providers/aws/events/alexa-smart-home)
+<!-- DOCS-SITE-LINK:END -->
+
+# Alexa Smart Home
+
+## Event definition
+
+This will enable your Lambda function to be called by an Alexa Smart Home Skill.
+`amzn1.ask.skill.xx-xx-xx-xx` is an application ID for Alexa Smart Home. You need to sign up [Amazon Developer  Console](https://developer.amazon.com/) and get your application ID.
```
Comment (pmuens) [inline docs/providers/aws/events/alexa-smart-home.md]: `Amazon Developer  Console` --> `Amazon Developer Console` (one space can be removed 😬).

File: docs/providers/aws/guide/resources.md [hunk 2]
```diff
@@ -70,7 +70,7 @@ We're also using the term `normalizedName` or similar terms in this guide. This
...
...daPermissionIotTopicRule{index} </li><li>**S3**: {normalizedFunctionName}LambdaPermission{normalizedBucketName}S3</li><li>**APIG**: {normalizedFunctionName}LambdaPermissionApiGateway</li><li>**SNS**: {normalizedFunctionName}LambdaPermission{normalizedTopicName}SNS</li><li>**Alexa Skill**: {normalizedFunctionName}LambdaPermissionAlexaSkill</li><li>**Alexa Smart Home**: {normalizedFunctionName}LambdaPermissionAlexaSmartHome{index}</li><li>**Cognito User Pool Trigger Source**: {normalizedFunctionName}LambdaPermissionCognitoUserPool{normalizedPoolId}TriggerSource{triggerSource}</li> </ul> | <ul><li>**Schedule**: HelloLambdaPermissionEventsRuleSchedule1</li><li>**CloudWatch Event**: HelloLambdaPermissionEventsRuleCloudWatchEvent1</li><li>**CloudWatch Log**: HelloLambdaPermissionLogsSubscriptionFilterCloudWatchLog1</li><li>**IoT**: HelloLambdaPermissionIotTopicRule1 </li><li>**S3**: HelloLambdaPermissionBucketS3</li><li>**APIG**: HelloLambdaPermissionApiGateway</li><li>**SNS**: HelloLambdaPermissionTopicSNS</li><li>**Alexa Skill**: HelloLambdaPermissionAlexaSkill</li><li>**Cognito User Pool Trigger Source**: HelloLambdaPermissionCognitoUserPoolMyPoolTriggerSourceCustomMessage</li> </ul>|
```
Comment (pmuens) [inline docs/providers/aws/guide/resources.md]: Could you add an example for the `Alexa Smart Home` resource name here (in the table all the examples are in the right column).

File: lib/plugins/aws/package/compile/events/alexaSmartHome/index.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            const lambdaPermissionLogicalId = this.provider.naming
+              .getLambdaAlexaSmartHomePermissionLogicalId(functionName,
+                alexaSmartHomeNumberInFunction);
+
+            const permissionCloudForamtionResource = {
+              [lambdaPermissionLogicalId]: permissionTemplate,
+            };
+
+            _.merge(this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
+              permissionCloudForamtionResource);
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/alexaSmartHome/index.js]: `permissionCloudForamtionResource` --> `permissionCloudFormationResource`

File: lib/plugins/aws/package/compile/events/alexaSmartHome/index.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                Principal: 'alexa-connectedhome.amazon.com',
+                EventSourceToken: EventSourceToken.replace(/\\n|\\r/g, ''),
+              },
+            };
+
+            const lambdaPermissionLogicalId = this.provider.naming
+              .getLambdaAlexaSmartHomePermissionLogicalId(functionName,
+                alexaSmartHomeNumberInFunction);
+
+            const permissionCloudForamtionResource = {
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/alexaSmartHome/index.js]: `permissionCloudForamtionResource` --> `permissionCloudFormationResource`

File: lib/plugins/aws/package/compile/events/alexaSmartHome/index.test.js [hunk 5]
```diff
@@ -0,0 +1,221 @@
+'use strict';
+
+const expect = require('chai').expect;
+const AwsProvider = require('../../../../provider/awsProvider');
+const AwsCompileAlexaSmartHomeEvents = require('./index');
+const Serverless = require('../../../../../../Serverless');
+
+describe('awsCompileAlexaSmartHomeEvents', () => {
```
Comment (pmuens) [inline lib/plugins/aws/package/compile/events/alexaSmartHome/index.test.js]: `AwsCompileAlexaSmartHomeEvents` (the plugins class name)

**Comment by horike37** [comment]
@pmuens 
fixed what you pointed out and added the example stuff to `Kotlin ` template :smile:
Thank you for giving us the notice @jvsinclair. Added that to the doc :muscle:

**Comment by jvsinclair** [comment]
@pmuens are you ok with his changes? I would be great to get this merged.

**Comment by pmuens** [comment]
Thanks for pinging @jvsinclair 👍

I'll do a final review the upcoming days!

File: docs/providers/aws/guide/resources.md [hunk 6]
```diff
@@ -70,7 +70,7 @@ We're also using the term `normalizedName` or similar terms in this guide. This
...
...onName}LambdaPermission{normalizedBucketName}S3</li><li>**APIG**: {normalizedFunctionName}LambdaPermissionApiGateway</li><li>**SNS**: {normalizedFunctionName}LambdaPermission{normalizedTopicName}SNS</li><li>**Alexa Skill**: {normalizedFunctionName}LambdaPermissionAlexaSkill</li><li>**Alexa Smart Home**: {normalizedFunctionName}LambdaPermissionAlexaSmartHome{index}</li><li>**Cognito User Pool Trigger Source**: {normalizedFunctionName}LambdaPermissionCognitoUserPool{normalizedPoolId}TriggerSource{triggerSource}</li> </ul> | <ul><li>**Schedule**: HelloLambdaPermissionEventsRuleSchedule1</li><li>**CloudWatch Event**: HelloLambdaPermissionEventsRuleCloudWatchEvent1</li><li>**CloudWatch Log**: HelloLambdaPermissionLogsSubscriptionFilterCloudWatchLog1</li><li>**IoT**: HelloLambdaPermissionIotTopicRule1 </li><li>**S3**: HelloLambdaPermissionBucketS3</li><li>**APIG**: HelloLambdaPermissionApiGateway</li><li>**SNS**: HelloLambdaPermissionTopicSNS</li><li>**Alexa Skill**: HelloLambdaPermissionAlexaSkill</li><li>**Alexa Smart Home**: HelloLambdaPermissionAlexaSmartHome</li><li>**Cognito User Pool Trigger Source**: HelloLambdaPermissionCognitoUserPoolMyPoolTriggerSourceCustomMessage</li> </ul>|
```
Comment (pmuens) [inline docs/providers/aws/guide/resources.md]: It looks like the example is missing an index entry at the end

**Comment by horike37** [comment]
@pmuens 
Just fixed the example section!

**Comment by pmuens** [comment]
Great! Thanks for updating @horike37 👍

**Pull request merged**